Wall time checkpoint

pbehne opened the pull request:
## Automatic checkpoints based on wall time

### Overview:
This pull request aims to enhance the checkpoint system. The changes encompass several key improvements, including introducing wall-time based checkpoints, standardizing checkpoint file structures, addressing logic bugs, refining parameter names for clarity, and optimizing logic for efficiency. Additionally, this PR includes comprehensive testing to ensure the reliability and functionality of the updated checkpoint system.

### Changes Included:
1. **Wall Time Checkpoints and Code Cleanup:**
   - Implemented automatic wall time checkpoints, allowing checkpoints to be written based on wall time intervals. This behavior can be modified through the new `Outputs/Checkpoint/wall_time_interval` input parameter.
   - Refactored common code from the child `Checkpoint` class into the parent `Output` class, enhancing code maintainability.
   - Renamed variables and deprecated input parameter names for improved clarity and consistency across tests. `interval` was renamed to `time_step_interval` and `minimum_time_interval` was renamed to `min_simulation_time_interval`.
   - Renamed the CLI parameter `--half-transient` to `--test-checkpoint-half-transient`.

2. **Standardizing Checkpoint File Structure:**
   - Modified the signal-based checkpoint folder naming convention to align with the user-specified checkpoint folder naming convection. This change enhances consistency and simplifies management.

3. **Logic Bug Patch:**
   - Patched a logic bug related to the update of `_last_output_time` (now renamed to `_last_output_simulation_time`), ensuring the variable is updated when it is supposed to.
   - Patched a bug that accessed `FileOutput::_file_base` before it was set to it's final (correct) value. MOOSE is now able to find the correct checkpoint directory when recovering a simulation.

4. **Override Default Time Step Interval:**
   - Enabled the override of the default `time_step_interval` when the `wall_time_interval` is specified but not the `time_step_interval`, ensuring flexibility in intended output frequency.

5. **Cleanup and Optimization:**
   - Removed the `delete_output_folder` parameter from `RunApp` as it was unused.
   - Optimized logic in `SignalTester`, particularly in `send_signal`, to improve efficiency and reliability during testing.
   - Only one checkpoint may now be defined per multiapp level.

6. **Test Coverage Enhancement:**
   - Added comprehensive test cases to validate the functionality of wall time checkpoints and to detect multiple checkpoint objects at a given multi-app level.
   - Introduced a command-line parameter (`output-wall-time-interval`) and corresponding code to facilitate automatic wall-time checkpoint testing without explicitly defining checkpoints in the input file.
   - Renamed the --half-transient command line parameter to --test-checkpoint-half-transient.

### Impact:
These changes significantly improve the robustness, flexibility, and maintainability of the checkpoint system. By addressing logic bugs, introducing new features, and refining existing functionalities, this PR lays the groundwork for a more efficient and reliable checkpoint system.

### Related PRs:
Bison:  https://github.inl.gov/ncrc/bison/pull/5803 (merged)
Griffin: https://github.inl.gov/ncrc/griffin/pull/1723 (merged), https://github.inl.gov/ncrc/griffin/pull/1767 (merged)

### References:
- Closes #13562

pbehne committed:
Initial step towards reworking Checkpoint system.

This commit modifies the signal-based checkpoint file structure to
use the same naming convention as the user-specified checkpoint file structure.
These changes were tested using the test_signal and test_signal_parallel tests
in tests/misc/signal_handler. Ref #26588.

pbehne committed:
Updated variable/type names to make more sense.

pbehne committed:
Patched logic bug for update of _last_output_time.

Output objects have an attribute _last_output_time describing the simulation
time at which output last occured. Before this commit, the updating of this
variable occurred before checking the status of shouldOutput, resulting in
erroneous variable updates when shouldOutput is false. This commit moves the
variable update into the check of shouldOutput, correcting the issue.

pbehne committed:
Added wall time checkpoint, code cleanup

Added the capability that a checkpoint is written every 10 minutes of wall time
that a simulation runs (this wall time interval can be changed through the new
“wall_time_interval” input parameter).

Additionally, common code from the parent Output object that was duplicated in
the child Checkpoint object was refactored into Output.

Finally, some variable names were renamed for improved clarity.
The “interval” input parameter was deprecated in favor of “time_step_interval”
and the “minimum_time_interval” input parameter was deprecated in favor of
“min_simulation_time_interval”.
These parameter names have correspondingly been updated in all tests.

pbehne committed:
Override default time_step_interval, precise timer

The time_step_interval input parameter has a default value of 1, i.e., output
is performed at every time step. In the event that the user explicitly specifies
the wall_time_interval parameter but not the time_step_interval parameter, it is
desirable that output does not occur at every time step. This commit sets
overrides the default value of time_step_interval from 1 to (practically)
infinity when the user specifies wall_time_interval.

pbehne committed:
Added tests for wall time checkpoints

This commit adds test cases for the new wall time checkpoint functionality.
To test that wall time checkpoints work without the user creating a checkpoint
block or using the shortcut syntax, the command line parameter
‘output-wall-time-interval’ (and supporting code) was added to MooseApp as a
way to override the default wall time interval of 10 minutes without defining a
checkpoint in the input file. The 10 minute default value needed to be
overridden to a much shorter interval suitable for the test suit.

pbehne committed:
clang-format

pbehne committed:
Added documentation.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2008875/) on 72aeab1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26588/site)

This comment will be updated on new commits.

pbehne committed:
Restored accidentally deleted default value of sync_times.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2010409/) on 9d63938 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1df316</th>
      <th colspan=4>#26588 9d6393</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.10%</td>
      <td>85.12%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>98656</td>
      <td>98712</td>
      <td>+56</td>
      <td>81</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17273</td>
      <td>17258</td>
      <td>-15</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26588/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26588/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26588/coverage/xfem)

</details>



This comment will be updated on new commits.

pbehne committed:
Test for multiple checkpoint objects

Added logic to detect multiple checkpoint objects at a given multiapp level and
error if more than one is found. Added test to check this logic.

pbehne committed:
Cleaned up definition of valid 'execute_on' options in Checkpoint.

pbehne committed:
Corrected typos.

pbehne committed:
Removed 'delete_output_folder' parameter from RunApp because it is not used.

pbehne commented:
@milljm, @loganharbour:

The test harness currently has the ability to delete old output file before re-running tests. This helps to prevent false successes and is done through the `delete_output_before_running` parameter. However, this only applies to files that are directly involved in determining whether a test passes (i.e., the exodus file in Exodiff tests, the check_files in CheckFiles tests, etc). Currently, there is no check for old checkpoint folders to delete them before running a test. I would like to incorporate a new boolean test harness parameter that, if set to true, looks for and deletes all *_cp directories in the test directory. I was talking to @permcody about what the default behavior should be. I think I am leaning towards having checkpoint folders deleted by default and making the user responsible for setting the new parameter to false for tests that require the presence of checkpoints for restart/recover. Thoughts?

milljm commented:
Instead of a bool to delete wildcard data, can we instead do something where if the key is populated, the value becomes what gets deleted?

```pre
[test]
     type = Exodiff
     exodiff = 'recover_out.e'
     delete_checkpoints = checkpoint_directory
     delete_output_before_running = true
[]
```
will delete `recover_out.e` and directories named `checkpoint_directory`, but leave any directories/files named `*_cp` alone.

loganharbour commented:
I think that I would prefer instead that we keep a registry of files that are created during the run, and put that in a common output file that we can read. that way we don't need to keep adding things to delete to the harness, but simply need to read the result.

this would tie in with @dschwen's idea of keeping a pid file around. but with more info

pbehne commented:
> Instead of a bool to delete wildcard data, can we instead do something where if the key is populated, the value becomes what gets deleted?
> 
> ```
> [test]
>      type = Exodiff
>      exodiff = 'recover_out.e'
>      delete_checkpoints = checkpoint_directory
>      delete_output_before_running = true
> []
> ```
> 
> will delete `recover_out.e` and directories named `checkpoint_directory`, but leave any directories/files named `*_cp` alone.

@milljm  I had a similar idea. The only annoying part about it is that it would require changing A LOT of test files without an easy way to automate the change process (have to manually determine the cp folder name for each test, etc). However, the amount of pain required to implement the change does not mean that the change itself is a bad idea.

milljm commented:
We are already creating a `.previous_results.json` file, can we use that (for @dschwen work as well)? And if that's not possible, we should probably talk about all these 'registration files' that are populating information during a TestHarness run. Try to consolidate them...

pbehne commented:
> I think that I would prefer instead that we keep a registry of files that are created during the run, and put that in a common output file that we can read. that way we don't need to keep adding things to delete to the harness, but simply need to read the result.
> 
> this would tie in with @dschwen's idea of keeping a pid file around. but with more info

@loganharbour I like this idea because of it's robustness. The implementation will probably be a separate PR, though. Tagging issue #9304.

milljm commented:
> > Instead of a bool to delete wildcard data, can we instead do something where if the key is populated, the value becomes what gets deleted?
> > ```
> > [test]
> >      type = Exodiff
> >      exodiff = 'recover_out.e'
> >      delete_checkpoints = checkpoint_directory
> >      delete_output_before_running = true
> > []
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > will delete `recover_out.e` and directories named `checkpoint_directory`, but leave any directories/files named `*_cp` alone.
> 
> @milljm I had a similar idea. The only annoying part about it is that it would require changing A LOT of test files without an easy way to automate the change process (have to manually determine the cp folder name for each test, etc). However, the amount of pain required to implement the change does not mean that the change itself is a bad idea.

My main concern is with the hard-coded wild card regular expression having an unintentional side effect of deleting data we did not intend.

loganharbour commented:
> We are already creating a `.previous_test_results.json` file, can we use that (for @dschwen work as well)? And if that's not possible, we should probably talk about all these 'registration files' that are populating information during a TestHarness run. Try to consolidate them...

No, this won't have anything to do with the test harness

milljm commented:
> > We are already creating a `.previous_test_results.json` file, can we use that (for @dschwen work as well)? And if that's not possible, we should probably talk about all these 'registration files' that are populating information during a TestHarness run. Try to consolidate them...
> 
> No, this won't have anything to do with the test harness

Sorry, no to what? We're not going to use `.previous_test_results.json`, or `moose_test.yaml`. That leaves us with Daniel's registry methods. I am fine with that. But won't this start to have things to do with the test harness? My concern is that that makes 3 different "registry files" that control TestHarness behavior.

pbehne committed:
Cleaned up logic in SignalTester.

The main part of this commit was reworking the logic in SignalTester.send_signal.
Namely, the check to see if the moose binary has started running and writing to the
output file was modified from an approach that copied and read the output file to
a much cheaper approach that simply queried the current position in the output file.

Additionally, the 'signal' parameter previously did not do anything. This has been patched.

Finally, the 'sleep_time' parameter has been removed because it was not deemed necessary to
the testing process (there should be no such thing as a signal being sent too early after the
binary has started to run; the checkpoint will simply be output at the end of the next time step).

pbehne commented:
@loganharbour, @permcody: This PR affects tests in Bison and Griffin (see links in PR description). I have draft PRs up for these apps and all tests are passing, but I have not yet updated deprecated parameter names there. I would like reviews and approvals here before doing that.

permcody commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -97,6 +97,13 @@ Checkpoint::Checkpoint(const InputParameters & parameters)
+  // If it is not used, _wall_time_interval keeps its current value.
+  // 'The --output-wall-time-interval parameter is necessary for testing
+  // and should only be used in the test suite.
+  Output::setWallTimeIntervalFromCommandLineParam();
```
Ah, I see your documentation here... It still might be helpful to mention the testing in the documentation next to the function declaration.

permcody reviewed (changes requested):
I left you a few comments. The only one that's a showstopper is the enum option list since it appears incorrect based on the documentation.

pbehne commented on framework/include/outputs/Checkpoint.h:
```diff
@@ -16,16 +16,20 @@
- * MODIFIED_EXISTING: We took an existing checkpoint and enabled autosave checks on it.
- * SYSTEM_AUTOSAVE: These checkpoints run in the background and output only when sent a signal.
+ * Enumerated type for determining what type of checkpoint this is.
+ * USER_CREATED: Checkpoint is requested by the user in the input file.
```
Good catch, will remove.

pbehne commented on framework/include/outputs/Checkpoint.h:
```diff
@@ -16,16 +16,20 @@
+ * SYSTEM_CREATED: This type of checkpoint is created automatically by the
+ *   system for the purpose of writing checkpoints at regularly scheduled
+ *   walltime intervals or when sent a signal.
+ * USER_AND_SYSTEM_CREATED: Checkpoint is requested by the user in the input
```
Yes, a checkpoint object is always created. Do you think "USER_CREATED" would be a better name, since the checkpoint is explicitly requested by the user and the system piggybacks off that object to write wall time and signal checkpoints?

pbehne commented on framework/include/outputs/Output.h:
```diff
@@ -165,6 +167,13 @@ class Output : public MooseObject,
+   * @param cli_param_name The name of the command line parameter to set the wall time interval to
+   *
+   */
+  void setWallTimeIntervalFromCommandLineParam();
```
Adding clarification.

loganharbour replied:
Still trying to think if there's a better way to do it. we typically try to avoid adding these kinds of methods when they're only used for testing

pbehne commented on framework/src/actions/AutoCheckpointAction.C:
```diff
@@ -25,23 +25,38 @@ AutoCheckpointAction::AutoCheckpointAction(const InputParameters & params) : Act
+  }
+
+  // We don't want to set up automatic checkpoints if we are not in the master app
+  else if (!_app.isUltimateMaster())
```
Enclosing in if statement.

pbehne commented on framework/src/base/MooseApp.C:
```diff
@@ -693,6 +700,13 @@ MooseApp::setupOptions()
 
   _half_transient = getParam<bool>("half_transient");
 
+  if (isParamValid("output_wall_time_interval"))
```
Are you sure about the name? I don't see anything called `addCheckedParam`.

permcody replied:
`addRangeCheckedParam` https://github.com/idaholab/moose/blob/1fbb5aad81de244e073f9e5ae406701ac6831edc/framework/include/utils/InputParameters.h#L174-L194

pbehne replied:
This cannot be used to replace a command line param, per our in-person conversation.

pbehne committed:
Addressed Cody's comments.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2033195/) on 8bcf005 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26588/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26588/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 796cb8aa68a15d61eb82eb353bb6582b01c6634f`

pbehne committed:
clang-format.

pbehne commented:
@permcody, @loganharbour: Ready to merge?

loganharbour commented on framework/include/base/MooseApp.h:
```diff
@@ -521,6 +521,29 @@ class MooseApp : public ConsoleStreamInterface,
+   * is valid)
+   */
+  template <typename T>
+  bool setVariableToCommandLineParam(T & variable_to_set, const std::string & cli_param_name) const
```
I think this is a very specific use case that doesn't necessitate a new method for it. Can you just put this logic into the method in `Console` instead? You can access `MooseApp::isParamValid()` from the `Console`

pbehne replied:
Did you mean `Output` (not `Console`)?

loganharbour replied:
correct

loganharbour replied:
So this no longer exists, right?

loganharbour commented on framework/src/actions/AutoCheckpointAction.C:
```diff
@@ -25,23 +25,39 @@ AutoCheckpointAction::AutoCheckpointAction(const InputParameters & params) : Act
+  {
+    // Get most recently added Checkpoint object and error
+    checkpoints.back()->paramError(
+        "_type",
```
Is this even a valid parameter to error on?

pbehne replied:
This is the param that makes the most sense to error on, since `type` will always be `Checkpoint`. Otherwise, I could just change it to a MOOSE error.

loganharbour replied:
`_type` is a private parameter and you'll never get a line reference to it (doesn't exist in input). Just make it a moose error

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -693,6 +700,13 @@ MooseApp::setupOptions()
+  {
+    const auto output_wall_time_interval = getParam<Real>("output_wall_time_interval");
+    if (output_wall_time_interval <= 0)
+      mooseError("output-wall-time-interval must be greater than zero.");
```
```suggestion
      mooseError("--output-wall-time-interval must be greater than zero.");
```

loganharbour commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -119,61 +126,48 @@ Checkpoint::directory() const
-  }
-
-  _current_execute_flag = EXEC_NONE;
+  FileOutput::outputStep(type);
```
Remove the override for `Checkpoint::outputStep()` as it just calls the parents method and nothing else

loganharbour commented on framework/src/outputs/Checkpoint.C:
```diff
@@ -119,61 +126,48 @@ Checkpoint::directory() const
-      // Reset signal number since we output
-      Moose::interrupt_signal_number = 0;
-    }
+    _console << "Unix signal SIGUSR1 detected. Outputting checkpoint file. \n";
```
```suggestion
    _console << "Unix signal SIGUSR1 detected. Outputting checkpoint file.\n";
```

loganharbour commented on framework/src/outputs/Output.C:
```diff
@@ -40,10 +40,28 @@ Output::validParams()
   // Output intervals and timing
-  params.addParam<unsigned int>(
-      "interval", 1, "The interval at which time steps are output to the solution file");
+  params.addParam<unsigned int>("time_step_interval",
```
Make this a range checked param (>0 or >=0)

loganharbour commented on framework/src/outputs/Output.C:
```diff
@@ -272,14 +309,39 @@ Output::onInterval()
+    output = false;
+
+  // check if enough wall time has passed between outputs
+  const auto now = std::chrono::steady_clock::now();
```
I think that we need to synchronize these times. Or let rank 0 decide and then communicate its choice to all procs. If we hit this at different times on different processors, we could take different branches in this object and it would really screw things up

pbehne replied:
Nice catch

loganharbour replied:
Shouldn't we do min?

pbehne replied:
The purpose of the wall time checkpoint feature can be oversimplified to the following statement: "Output a checkpoint every 10 minutes of wall time a simulation runs."

On line 323 below, each processor computes the wall time since the last output was written. One processor will have the smallest wall time since last output and one will have the largest wall time since last output. I could see it working both ways (using min or max), but I'd argue that using max is the way to go since with using min, it could be longer than 10 minutes before the next output. In practice, it probably doesn't matter too much, so I'm willing to change it to min if you have an argument for it.

loganharbour commented on modules/navier_stokes/test/tests/finite_volume/ins/channel-flow/tests:
```diff
@@ -23,6 +23,7 @@
     method = "!dbg"
     requirement = 'The system shall be able to model free-slip conditions in a channel using the NSFV action.'
     valgrind = HEAVY
+    recover = false
```
Why?

pbehne replied:
recovery is manually tested in `[checkpoint-restart]` below.

loganharbour commented on python/TestHarness/testers/SignalTester.py:
```diff
@@ -21,39 +19,62 @@ class SignalTester(RunApp):
 
-    def send_signal(self,pid):
-        """Function used to send a signal to the program automatically for testing purposes."""
+        valid_signals = {
+            "SIGUSR1" : signal.SIGUSR1,
+            "SIG_DFL" : signal.SIG_DFL,
+            "SIG_IGN" : signal.SIG_IGN,
+            "SIGABRT" : signal.SIGABRT,
+            "SIGBUS"  : signal.SIGBUS,
+            "SIGCHLD" : signal.SIGCHLD,
+            "SIGCONT" : signal.SIGCONT,
+            "SIGFPE"  : signal.SIGFPE,
+            "SIGHUP"  : signal.SIGHUP,
+            "SIGILL"  : signal.SIGILL,
+            "SIGINT"  : signal.SIGINT,
+            "SIGKILL" : signal.SIGKILL,
+            "SIGPIPE" : signal.SIGPIPE,
+            "SIGSEGV" : signal.SIGSEGV,
+            "SIGTERM" : signal.SIGTERM,
+            "SIGUSR2" : signal.SIGUSR2,
+            "SIGWINCH": signal.SIGWINCH
+        }
```
```suggestion
        valid_signals = {}
        for sig in signal.Signals:
            valid_signals[sig.name] = sig
```

loganharbour commented on test/tests/outputs/checkpoint/tests:
```diff
@@ -66,7 +65,9 @@
                   "format:"
+    issues = '#1927 #13562'
     [recover_half_transient]
+      #issues = '#1927'
```
```suggestion
```

loganharbour commented on test/tests/outputs/checkpoint/tests:
```diff
@@ -105,6 +106,7 @@
     []
     [recover]
       # Gold for this test was created using checkpoint.i without any recover options
+      #issues = '#1927'
```
```suggestion
```

loganharbour commented on test/tests/outputs/checkpoint/tests:
```diff
@@ -117,13 +119,45 @@
       detail = "be capable of restarting a simulation from the output data."
     []
+
+    [wall_time_interval]
+      #issues = '#13562'
+      type = RunApp
+      input = checkpoint.i
+      recover = false
+      cli_args = 'Outputs/file_base=checkpoint_wall_time_interval_out --output-wall-time-interval 0.02'
+      detail = "at specified wall time intervals"
+    []
+    [wall_time_interval_recover]
+      #issues = '#13562'
+      type = Exodiff
+      input = checkpoint.i
+      exodiff = checkpoint_wall_time_interval_out.e
+      recover = false
+      prereq = 'default/wall_time_interval'
+      delete_output_before_running = false
+      cli_args = 'Outputs/file_base=checkpoint_wall_time_interval_out --recover'
+      expect_out = 'Using checkpoint_wall_time_interval_out_cp/\d+ for recovery'
+      detail = "be capable of restarting a simulation from the wall time checkpoint."
+    []
+
+    [output_wall_time_interval_error]
+      #issues = '#13562'
+      type = 'RunException'
+      input = 'checkpoint.i'
+      cli_args = '--output-wall-time-interval -1'
+      expect_err = 'output-wall-time-interval must be greater than zero.'
+      detail = ', catching invalid values of output-wall-time-interval.'
+    []
```
```suggestion
    [wall_time_interval]
      type = RunApp
      input = checkpoint.i
      recover = false
      cli_args = 'Outputs/file_base=checkpoint_wall_time_interval_out --output-wall-time-interval 0.02'
      detail = "at specified wall time intervals"
    []
    [wall_time_interval_recover]
      type = Exodiff
      input = checkpoint.i
      exodiff = checkpoint_wall_time_interval_out.e
      recover = false
      prereq = 'default/wall_time_interval'
      delete_output_before_running = false
      cli_args = 'Outputs/file_base=checkpoint_wall_time_interval_out --recover'
      expect_out = 'Using checkpoint_wall_time_interval_out_cp/\d+ for recovery'
      detail = "be capable of restarting a simulation from the wall time checkpoint."
    []

    [output_wall_time_interval_error]
      type = 'RunException'
      input = 'checkpoint.i'
      cli_args = '--output-wall-time-interval -1'
      expect_err = 'output-wall-time-interval must be greater than zero.'
      detail = ', catching invalid values of output-wall-time-interval.'
    []
```

loganharbour commented on test/tests/outputs/checkpoint/tests:
```diff
@@ -117,13 +119,45 @@
+    issues = '#1927 #13562 #25755'
     [recover_with_checkpoint_block_half_transient]
       # Tests that recover works when defining as a Output sub-block
+      #issues = '#1927'
```
```suggestion
```

loganharbour commented on test/tests/outputs/checkpoint/tests:
```diff
@@ -161,6 +195,7 @@
     []
     [recover_with_checkpoint_block]
       # Gold for this test was created using checkpoint_block.i without any recover options
+      #issues = '#1927'
```
```suggestion
```

loganharbour commented on test/tests/outputs/checkpoint/tests:
```diff
@@ -173,13 +208,43 @@
       detail = "be capable of restarting a simulation from the output data."
     []
+
+    [wall_time_interval]
+      #issues = '#13562'
+      type = RunApp
+      input = checkpoint_block.i
+      recover = false
+      cli_args = 'Outputs/file_base=checkpoint_block_wall_time_interval_out --output-wall-time-interval 0.02'
+      detail = "at specified wall time intervals"
+    []
+    [wall_time_interval_recover]
+      #issues = '#13562'
+      type = Exodiff
+      input = checkpoint_block.i
+      exodiff = checkpoint_block_wall_time_interval_out.e
+      recover = false
+      prereq = 'block/wall_time_interval'
+      delete_output_before_running = false
+      cli_args = 'Outputs/file_base=checkpoint_block_wall_time_interval_out --recover'
+      expect_out = 'Using checkpoint_block_wall_time_interval_out_cp/\d+ for recovery'
+      detail = "be capable of restarting a simulation from the wall time checkpoint."
+    []
+
```
```suggestion
    [wall_time_interval]
      type = RunApp
      input = checkpoint_block.i
      recover = false
      cli_args = 'Outputs/file_base=checkpoint_block_wall_time_interval_out --output-wall-time-interval 0.02'
      detail = "at specified wall time intervals"
    []
    [wall_time_interval_recover]
      type = Exodiff
      input = checkpoint_block.i
      exodiff = checkpoint_block_wall_time_interval_out.e
      recover = false
      prereq = 'block/wall_time_interval'
      delete_output_before_running = false
      cli_args = 'Outputs/file_base=checkpoint_block_wall_time_interval_out --recover'
      expect_out = 'Using checkpoint_block_wall_time_interval_out_cp/\d+ for recovery'
      detail = "be capable of restarting a simulation from the wall time checkpoint."
    []

```

loganharbour commented on test/tests/outputs/checkpoint/tests:
```diff
@@ -173,13 +208,43 @@
+    []
+
     [execute_on_param_error]
+      #issues = '#25755'
```
```suggestion
```

loganharbour commented on test/tests/outputs/recover/tests:
```diff
@@ -8,8 +8,8 @@
                   'test_recover_dir_cp/0005-restart-0.rd/header'
-    delete_output_folders = false
     cli_args = 'Executioner/num_steps=5'
+    recover = false
```
Why?

pbehne replied:
When recovery is tested, the test harness appends `Outputs/checkpoint=true` to the command used to run the input. If the input already has a checkpoint block defined in it, recovery testing is basically asking MOOSE to create a second `Checkpoint` object, which is no longer allowed (since everything can be accomplished with one object). Furthermore, if the author of a test puts a checkpoint block in their input, it seems reasonable to assume that they also add a test that recovers/restarts from that checkpoint, making it pointless to run the test in recovery mode.

loganharbour commented on test/tests/postprocessors/pps_interval/pps_out_interval.i:
```diff
@@ -129,10 +129,15 @@
   [../]
 []
+
+[Debug]
+  show_parser = true
+  show_actions = true
+[]
```
Why?

pbehne replied:
Because I made an oppsie.

loganharbour commented on test/tests/restart/kernel_restartable/tests:
```diff
@@ -10,6 +10,7 @@
       input = 'kernel_restartable.i'
       exodiff = 'kernel_restartable_out.e'
       detail = "perform residual calculations that accumulate state and"
+      recover = false
```
Why?

pbehne replied:
Restart is already tested below in `[test2]`.

loganharbour commented on test/tests/restart/kernel_restartable/tests:
```diff
@@ -29,6 +30,7 @@
       input = 'kernel_restartable_custom_name.i'
       exodiff = 'kernel_restartable_custom_name_out.e'
       detail = "to name problem objects in a simulation and"
+      recover = false
```
Why?

pbehne replied:
Restart is already tested below in `[custom_name2]`.

loganharbour commented on test/tests/time_steppers/iteration_adaptive/tests:
```diff
@@ -50,6 +50,7 @@
     csvdiff = 'adapt_tstep_grow_dtfunc_out.csv'
     requirement = "The system shall support the ability to grow the time step size
     when specifying the values of t and dt in the TimeStepper."
+    recover = false
```
Why?

pbehne replied:
Restart is already tested below in `[adapt_tstep_grow_dtfunc_restart]`.

pbehne commented on test/tests/outputs/recover/tests:
```diff
@@ -22,7 +22,6 @@
     cli_args = '--recover test_recover_dir_cp/0005'
```
See, recovery is tested here anyway.

pbehne committed:
Addressed Logan's review.

This included a bug fix to ensure _wall_time_since_last_output is the same
accross all processors. The --half-transient flag was renamed to
--test-checkpoint-half-transinet, and logic was added to allow tests with
explicitly created checkpoints to run in recover mode.

pbehne committed:
clang format

pbehne committed:
Fixed non-working logic in SignalTester.

pbehne committed:
Changed old --half-transient to new --test-checkpoint-half-transient.

pbehne committed:
Fixed non-working logic in SignalTester... again...

pbehne committed:
Patched bug causing seg fault

After adding a system-created checkpoint in AutoCheckpointAction, the checkpoints
vector does not know about this new chackpoint, resulting in a segfault when trying
to access the newly created checkpoint. This patch gets the new checkpoint from
the OutputWarehouse.

Also, a pass was changed to continue in SignalTester.

pbehne committed:
Patched checkpoint directory naming bug

Moose no longer allows multiple checkpoint objects. Because of this, input files
with user-specified checkpoints are no longer overwritten with the shortcut
`Outputs/checkpoint=true` when using the --test_checkpoint-half-transient flag
(previously known as the --half-transient flag). This changed enabled the
discovery of a bug when running moose in --recover mode: with checkpoint objects
named anything other than `out` moose was unable to find the checkpoint directory.
This is because moose was getting an empty value for the `file_base` parameter
in the checkpoint object before it was later set during the execution of the
Executioner. This commit patches this issue, ensuring the correct checkpoint
directory name is found.

pbehne committed:
Added checkpoint directory naming case where file_base is set by user.

pbehne committed:
In the last commit, a deleted like was restored to the wrong place. This commit moves it to the right place.

pbehne commented:
@loganharbour, Griffin calls `_app.halfTransient`, which has been renamed to `_app.testCheckpointHalfTransient`. We need to break next to get this Griffin PR in: https://github.inl.gov/ncrc/griffin/pull/1767

pbehne committed:
Added deprecation for MooseApp::halfTransient.

pbehne commented:
> @loganharbour, Griffin calls `_app.halfTransient`, which has been renamed to `_app.testCheckpointHalfTransient`. We need to break next to get this Griffin PR in: https://github.inl.gov/ncrc/griffin/pull/1767

Never mind, I added a deprecation option for MooseApp::halfTransient, so we should be able to pass griffin tests now.

pbehne commented:
@loganharbour, it is ready.

loganharbour commented on test/tests/outputs/recover/tests:
```diff
@@ -8,7 +8,6 @@
                   'test_recover_dir_cp/0005-mesh.cpr/meta_data_mesh.rd/header '
                   'test_recover_dir_cp/0005-restart-0.rd/data '
                   'test_recover_dir_cp/0005-restart-0.rd/header'
-    delete_output_folders = false
```
why? same with the other two

loganharbour commented on framework/include/base/MooseApp.h:
```diff
@@ -519,7 +519,13 @@ class MooseApp : public ConsoleStreamInterface,
    */
-  bool halfTransient() const { return _half_transient; }
+  bool testCheckpointHalfTransient() const { return _test_checkpoint_half_transient; }
+  bool halfTransient() const
```
What apps use this method? Just griffin?

pbehne replied:
Yup! I'll make another PR to remove the deprecated option once I update Griffin.

loganharbour commented:
Please resolve the conversations that you explicitly followed and made changes for. This looks good

Let's check with cody if `--test-checkpoint-half-transient` is too long, but I think I'm fine with it.

pbehne commented:
> Please resolve the conversations that you explicitly followed and made changes for. This looks good
> 
> Let's check with cody if `--test-checkpoint-half-transient` is too long, but I think I'm fine with it.

@permcody, see above.

pbehne committed:
Changed paramError to mooseError.

loganharbour commented:
Just need to rebase and deal with the conflicts. Good to go after.

pbehne commented:
Should be good now.

loganharbour reviewed (approved):
Added more tests since this is a pretty significant change

roystgnr commented:
Any chance the changes here are responsible for the new libMesh MOOSE-test CI failures?  E.g. https://civet.inl.gov/job/2127543/ - 3 of our recipes are giving "unknown dependency" failures from the non-skipped `misc/signal_handler.test_signal_*recover*` tests.

GiudGiud commented:
only thing is common is this `delete_output_before_running` parameter

pbehne commented:
I'll try to replicate the `signal_handler` failure in a container on rod and report back.

pbehne commented:
I spoke with @milljm, and he says there really isn't a way to easily replicate the environment used in the CIVET job. He can comment more on the specifics.

milljm commented:
I am looking a little more closely at this and I think this failing target was neglected to move over to use containers? I am seeing this target loading the following modules:
```pre
 5) moose-petsc-0b2deac-gcc-9.4.0-4wj7ppk <aL>      
 6) moose-petsc/0b2deac <aL>                        
 7) moose-petsc/libmesh_ci                          
 8) python-3.7.8-gcc-9.4.0-g
```
and I am scratching my head as to how this works. I don't see PETSc being built, so we must be using some old version indeed.

I think this target needs to be switched over to use our containers.

jwpeterson commented:
The full list of checks that are failing with this error is:
```
Test 32bit
Test MOOSE min clang
Test debug
```
They are all checks that run on libmesh PRs, so they might have been overlooked.

loganharbour commented:
I'm working on these slowly.

dschwen commented:
> and `minimum_time_interval` was renamed to `min_simulation_time_interval`.

Not in the `CommonOutputAction`, @pbehne !
